src/pybind/mgr/progress/module.py

```diff
@@ -0,0 +1,425 @@
+        super(PgRecoveryEvent, self).__init__(message, refs)
+
+        self._pgs = which_pgs
+
+        self._evacuate_osds = evactuate_osds
+
+        self._original_pg_count = len(self._pgs)
+
+        self._original_bytes_recovered = None
+
+        self._progress = 0.0
+
+        self.id = str(uuid.uuid4())
+
+    def pg_update(self, pg_dump, log):
+        # FIXME: O(pg_num) in python
+        # FIXME: far more fields getting pythonized than we really care about
+        pg_to_state = dict([(p['pgid'], p) for p in pg_dump['pg_stats']])
+
+        if self._original_bytes_recovered is None:
+            self._original_bytes_recovered = {}
+            for pg in self._pgs:
+                pg_str = str(pg)
+                log.debug(json.dumps(pg_to_state[pg_str], indent=2))
+                self._original_bytes_recovered[pg] = \
+                    pg_to_state[pg_str]['stat_sum']['num_bytes_recovered']
+
+        complete_accumulate = 0.0
+
+        # Calculating progress as the number of PGs recovered divided by the
+        # original where partially completed PGs count for something
+        # between 0.0-1.0. This is perhaps less faithful than lookign at the
+        # total number of bytes recovered, but it does a better job of
+        # representing the work still to do if there are a number of very
+        # few-bytes PGs that still need the housekeeping of their recovery
+        # to be done. This is subjective...
+
+        complete = set()
+        for pg in self._pgs:
+            pg_str = str(pg)
+            info = pg_to_state[pg_str]
+            state = info['state']
+
+            states = state.split("+")
+
+            unmoved = bool(set(self._evacuate_osds) & (
+                set(info['up']) | set(info['acting'])))
+
+            if "active" in states and "clean" in states and not unmoved:
+                complete.add(pg)
+            else:
+                # FIXME: handle apparent negative progress
+                #   (e.g. if num_bytes_recovered goes backwards)
+                # FIXME: handle apparent >1.0 progress
+                #   (num_bytes_recovered incremented by more than num_bytes)
+                if info['stat_sum']['num_bytes'] == 0:
+                    # Empty PGs are considered 0% done until they are
+                    # in the correct state.
+                    pass
+                else:
+                    recovered = info['stat_sum']['num_bytes_recovered']
+                    total_bytes = info['stat_sum']['num_bytes']
+                    ratio = float(recovered -
+                                  self._original_bytes_recovered[pg]) / \
+                        total_bytes
+
+                    complete_accumulate += ratio
+
+        self._pgs = list(set(self._pgs) ^ complete)
+        completed_pgs = self._original_pg_count - len(self._pgs)
+        self._progress = (completed_pgs + complete_accumulate)\
+            / self._original_pg_count
```

Comment:
nothing really pops out as obviously wrong in this commit. this final progress calculation looks strange to me because of the units. it seems to heavily (like really heavily) weight the completed pgs, treating the totality of all non-completed PGs as at most 1 PG in the final progress?

NM: i see now complete_accumulate is adding up per-PG complete ratios. OK looks good